### ISSUE TITLE
Command Palette: Moving away from cmdk

### ISSUE THREAD

**Issue by t-hamano** [WordPress/gutenberg#76950]
## What problem does this address?

The command palette depends on [cmdk](https://github.com/dip/cmdk), but we are encountering several accessibility issues.

- https://core.trac.wordpress.org/ticket/64923
- https://github.com/dip/cmdk/issues/373
- https://github.com/WordPress/gutenberg/issues/54634

As far as I know, these issues need to be resolved upstream, but unfortunately, it doesn't seem to be actively maintained.

Furthermore, it has been pointed out here that cmdk has accessibility issues.

https://github.com/WordPress/gutenberg/pull/49330#issuecomment-1645523624

## What is your proposed solution?

Technically, we should be able to completely decouple from cmdk and build our own fully accessible command palette. Since cmdk's core functionality is [contained within a single file](https://github.com/dip/cmdk/blob/dd2250ed608443e8f32bafc5fa2d1d07a3746aa3/cmdk/src/index.tsx), the amount of code required for implementation should not be excessive.

Please let me know if there are any obstacles to moving away from cmdk.

cc @WordPress/gutenberg-core @joedolson

**Comment by ramonjd** [comment]
Makes sense to me if the blocking issues are upstream.

What are the hard parts to replacing cmdk? Does `ComboboxControl` component solve any of the `aria/aria-activedescendant` patterns?

**Comment by joedolson** [comment]
I'm highly in favor of replacing it. We can fork it,  if necessary, but I suspect we can solve a lot of the problems with existing components.

**Comment by t-hamano** [comment]
> Does `ComboboxControl` component solve any of the `aria/aria-activedescendant` patterns? 

It seems that `ComboboxControl` does not use the `aria-activedescendant` pattern in the first place. Instead, we might be able to use a `Composite` component 🤔

**Comment by andrewserong** [comment]
Rolling our own makes a lot of sense to me, especially since we're using the command palette everywhere now.

**Comment by Mamaduka** [comment]
We could replace it and use Base UI components. I think that's what the new UI package is using under the hood, and it also includes a Command Palette example: https://base-ui.com/react/components/autocomplete#command-palette.

**Comment by t-hamano** [comment]
Thank you all for your feedback!

> a Command Palette example: https://base-ui.com/react/components/autocomplete#command-palette.

This is very interesting. This UI appears to be highly accessible.

Perhaps we could also refer to Ariakit's patterns: https://ariakit.com/examples/combobox-group

**Comment by jameskoster** [comment]
I suspect the `ScrollArea` and `AutoComplete` primitives used in the Base UI Command Palette example would be useful in other contexts too.

**Comment by youknowriad** [comment]
For me it's an implementation detail, I don't really have strong opinions about using it or removing it. Important part is fixing the bugs.

**Comment by t-hamano** [comment]
> We could replace it and use Base UI components. I think that's what the new UI package is using under the hood

If we adopt this approach, will we use @base-ui/react directly in the command palette, or will we need to wait for @wordpress/ui to be exposed?

**Comment by mirka** [comment]
> will we use @base-ui/react directly in the command palette, or will we need to wait for @wordpress/ui to be exposed?

@senadir and I were discussing this recently as well, and the plan is to use the `Autocomplete` primitives from `@wordpress/ui` (#74178). These will already have some styling that adheres to the design system, but be flexible enough to use in a command palette like this. And if it's not, we'll probably want to tweak the primitives themselves so they are flexible enough.

Opening up the `@wordpress/ui` package for general use has been a bit delayed due to some blockers (#76135) and minor priority shifts, but the visual and API design of our `Autocomplete` primitive is pretty much ready. We definitely want to have them available for the 7.1 development cycle.

**Comment by senadir** [comment]
As Lena said, the initial plan was to move to Base UI, which was delayed to 7.1
With `Autocomplete` component built-in now (based on Base UI's one), we would use that instead, part of the styling/UI updates we're doing to the component https://github.com/WordPress/gutenberg/issues/75616

**Comment by t-hamano** [comment]
I have been vibe-coding with the idea of rebuilding the command palette using the new Autocomplete component.

https://github.com/WordPress/gutenberg/compare/refactor-command-palette?expand=1

Fortunately, it seems we can maintain all features and design.

| Before | After |
|--------|--------|
| <img width="587" height="322" alt="Image" src="https://github.com/user-attachments/assets/d9e23380-ce33-4047-8bd7-3486374d3c0c" /> | <img width="599" height="324" alt="Image" src="https://github.com/user-attachments/assets/6883c665-1926-488a-828e-e6a2b43ae117" /> | 

However, the command palette has lists that are grouped, and each group has a label, which is not supported by the current `Autocomplete` component. To move this issue forward, we will likely need to add [Group](https://github.com/WordPress/gutenberg/compare/refactor-command-palette?expand=1#diff-e6d1210088a4283a2c01f3d1446ed14b3d7c2edeb8f053836e29b58c53eb9465) and [GroupLabel](https://github.com/WordPress/gutenberg/compare/refactor-command-palette?expand=1#diff-f1e7d80b616e3df48fab325a732ba22cf5fa47f1b0ac9481c4a87626de4183f7) components to the `Autocomplete` component. What do you think?

**Comment by mirka** [comment]
Ooh, thanks for testing it out! Good to know it will work.

> To move this issue forward, we will likely need to add [Group](https://github.com/WordPress/gutenberg/compare/refactor-command-palette?expand=1#diff-e6d1210088a4283a2c01f3d1446ed14b3d7c2edeb8f053836e29b58c53eb9465) and [GroupLabel](https://github.com/WordPress/gutenberg/compare/refactor-command-palette?expand=1#diff-f1e7d80b616e3df48fab325a732ba22cf5fa47f1b0ac9481c4a87626de4183f7) components to the `Autocomplete` component. What do you think?

Right, we'll be adding those subcomponents (and maybe some others) to the combobox/autocomplete/select category of primitives.

Another todo task on my list is to remove the `default` size for the `Item` subcomponent, defaulting to the `compact` size. But otherwise, I think we can proceed with implementing command palette with the `Autocomplete` primitives.
